### ISSUE TITLE
Consider assuming that untyped decorators pass through the signature unchanged

### ISSUE THREAD

**Issue by afonsotrepa** [astral-sh/ty#1971]
### Summary

I ran into this problem where ty wouldn't catch a mistyped return type but pyright would. Narrowed it down to a decorator I'm using.
Sorry if this is a known issue (a quick search didn't find another open issue for this) or limitation.

### Example
```Python
def deco(_):
    def inner(func):
        return func
    return inner


@deco(0)
def f() -> str:
    return ""


def g() -> float:
    return f()
```

Expected to get "error[invalid-return-type]" (which is indeed what we see if we remove the decorator) but instead got all checks passing.

### Version

0.0.2

P.S.: Still loving ty and ruff, keep up the good work!

**Comment by sharkdp** [comment]
Thank you for reporting this.

We currently don't support return type inference yet (which is why the return type of `deco` is just `Unknown`), but plan to add it soon. See #128 for details. 

Note: It's actually not completely clear to me how other type checkers handle this case. It seems to me like we would have to infer a generic `∀ T. Callable[[T], T]` type for `inner` here (and, in effect, for the return type of `deco`)? But pyright simply infers the type of `deco` as `(_: Unknown) -> ((func: Unknown) -> Unknown)`. So it looks like it might use additional heuristics (something like: "if the return type of a decorator is Unknown, just pretend it acts like an identity operation")?

**Comment by carljm** [comment]
Yes, I think I remember some discussion of this. At least pyright (maybe some other type checkers too) do make the assumption that an untyped decorator doesn't change the signature. I'm not sold on this idea, though I can see why it's attractive.

**Comment by carljm** [comment]
I'm not sure this is really blocked by #128, since I don't think #128 will solve it, nor is necessary to "solve" it in the way that other type checkers do.

**Comment by afonsotrepa** [comment]
Actually this issue is still present if we type hint the decorator (and this was my original issue, I was just trying to create a minimalist example but clearly cut too much out, sorry):

```Python
from collections.abc import Callable

def deco[**P, R](_: object) -> Callable[[Callable[P, R]], Callable[P, R]]:
    def inner(func: Callable[P, R]) -> Callable[P, R]:
        return func
    return inner


@deco(0)
def f() -> str:
    return ""


def g() -> float:
    return f()
```

Same deal, ty (0.0.2) says all checks passed while pyright (1.1.407) correctly detects a "reportReturnType" (Type "str" is not assignable to return type "float")  error.

**Comment by carljm** [comment]
@afonsotrepa Thanks for the example! It looks like that is a different limitation: https://github.com/astral-sh/ty/issues/1136

**Comment by carljm** [comment]
@afonsotrepa note also that this version works in all type checkers (including ty) today, because it doesn't rely on the special-cased heuristic in #1136:

```py
from typing import Callable, Protocol

class Identity(Protocol):
    def __call__[**P, R](self, fn: Callable[P, R], /) -> Callable[P, R]: ...

def deco(_: object) -> Identity:
    def inner[**P, R](func: Callable[P, R]) -> Callable[P, R]:
        return func
    return inner

@deco(0)
def f() -> str:
    return ""


def g() -> float:
    return f()
```

**Comment by afonsotrepa** [comment]
@carljm your Identity class fixed my issue, thanks!

**Comment by carljm** [comment]
We could also do this in a way that is LSP-specific, by implementing some form of "best-guess types" that we won't use to emit type diagnostics, but will use for LSP purposes.

**Comment by afonsotrepa** [comment]
@carljm sorry for another "off-topic" comment but I found your Identity class was not working for generic functions when using functools.wraps, so instead I'm now using:

```Python
def deco[**P, R](_: object) -> Callable[[Callable[P, R]], Callable[P, R]]:

    def dec(func: Callable[P, R]) -> Callable[P, R]:

        @functools.wraps(func)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
                return func(*args, **kwargs)

        return wrapper

    return dec
```

This works with ty and pyright and successfully catches invalid return types, even for functions with generics. Hopefully this will be useful for someone else too.

**Comment by rivershah** [comment]
Any updates please? `keras` and other ml frameworks UX is pretty impacted by this.

**Comment by carljm** [comment]
> Any updates please? `keras` and other ml frameworks UX is pretty impacted by this.

No updates; this hasn't made it to the top of the priority list.

The symptom you are seeing is in LSP ty doesn't know the types of decorated functions?

**Comment by rivershah** [comment]
Correct. The issue is a visual annoyance over anything functional. Libraries such as `keras` or `pytorch` are decorating nearly every fn. `pylance` gives proper function highlights whereas `ty` drops them losing familiar visual cues throughout the code base.

**Comment by carljm** [comment]
This would be very easy to implement, if we just make the assumption that an untyped decorator is an identity. The problem is that this assumption isn't particularly justified. It rests on an assumption that may be true for most decorators? But certainly is not true for all.

**Comment by rivershah** [comment]
Could we replicate behaviour similar to pyright/ pylance? Thanks

**Comment by carljm** [comment]
We should probably just see what this looks like in the ecosystem report and, if it looks good (which it probably will, since it's what other type checkers do), just do it, despite the unsoundness.

**Comment by charliermarsh** [comment]
I can try this since I worked on class decorators which has some similar logic.

**Comment by charliermarsh** [comment]
The ecosystem impact here is pretty dramatic: https://github.com/astral-sh/ruff/pull/26143

**Comment by AlexWaygood** [comment]
> The ecosystem impact here is pretty dramatic: [astral-sh/ruff#26143](https://github.com/astral-sh/ruff/pull/26143)

Note that the vast majority of diagnostics are on one project though:

<img width="1316" height="1349" alt="Image" src="https://github.com/user-attachments/assets/5784225b-01af-4186-8e4f-fd1180fb5043" />

What's sympy up to here?

**Comment by charliermarsh** [comment]
Yeah. It's still over 1000 new diagnostics on other projects though, not nothing! (Planning to look into it more before opening for review.)

**Comment by AlexWaygood** [comment]
> It's still over 1000 new diagnostics on other projects though, not nothing!

Right. Though really the _point_ of this change is so that users get more precise inferred types (aka, more diagnostics!), so _some_ fallout is expected. I agree this is on the larger end, though.

**Comment by charliermarsh** [comment]
Codex's analysis, omitting SymPy which are almost all false positives from one pattern:
```
Project            TP     FP    Total
━━━━━━━━━━━━━━━━  ━━━━━  ━━━━━  ━━━━━━━
SciPy              23    242      265
────────────────  ─────  ─────  ───────
Ibis               46    192      238
────────────────  ─────  ─────  ───────
Spack             163     57      220
────────────────  ─────  ─────  ───────
JAX                19    153      172
────────────────  ─────  ─────  ───────
scikit-learn       39     66      105
────────────────  ─────  ─────  ───────
dd-trace-py        27     70       97
────────────────  ─────  ─────  ───────
mitmproxy          32     33       65
────────────────  ─────  ─────  ───────
Open Library        2     50       52
────────────────  ─────  ─────  ───────
Manticore           4     39       43
────────────────  ─────  ─────  ───────
Torchvision        21     14       35
────────────────  ─────  ─────  ───────
Pandera             2     21       23
────────────────  ─────  ─────  ───────
cloud-init          0      5        5
────────────────  ─────  ─────  ───────
PaaSTA              0      4        4
────────────────  ─────  ─────  ───────
Bokeh               2      1        3
────────────────  ─────  ─────  ───────
pwndbg              2      0        2
────────────────  ─────  ─────  ───────
cki-lib             0      2        2
────────────────  ─────  ─────  ───────
Jinja               0      2        2
────────────────  ─────  ─────  ───────
speedrun.com        0      2        2
────────────────  ─────  ─────  ───────
zope.interface      0      2        2
────────────────  ─────  ─────  ───────
setuptools          0      1        1
```

I find these numbers not very compelling. I think doing this well requires some amount of return-type inference to avoid false positives. For example, given:

```python
def identity(func):
    return func

def replace(func):
    def inner(x: int) -> int:
        return x
    return inner


@identity
def preserved(x: int, y: str) -> str:
    return y


@replace
def replaced(x: int, y: str) -> str:
    return y
```

I believe Pyright actually looks at the return type of the decorator, to distinguish between returning `func` (which is `Unknown`) and returning `inner` which has a known type / signature (even though `replace` is untyped).

**Comment by carljm** [comment]
It looks like this (as described in this issue) isn't really what other type checkers do; that was a mistaken assumption on my part based on limited information. It seems like pyright and pyrefly both get types for some decorators via return-type inference. Pyright then additionally assumes that an untyped decorator returning `Unknown` (or returning a function with a signature like `(*args, **kwargs) -> ...`) is a pass-through decorator; pyrefly does not do this. Mypy does neither RTI nor makes assumptions about untyped decorators; it just goes with `Any` if the decorator is not typed; this is our current behavior.

I think that means we should defer this until after we have RTI and then evaluate the additional impact of making some assumptions similar to those pyright makes.

In the meantime, for anyone running into this issue, the best option is to see if your commonly used decorators could be explicitly typed (maybe in stubs, if third party) using `ParamSpec` to accurately describe their impact on the signature of the decorated function.

**Comment by carljm** [comment]
Thanks @charliermarsh for experimenting with this and collecting the ecosystem impact data!
